src/org/javarosa/core/model/condition/Constraint.java

```diff
@@ -70,7 +73,7 @@ public String getConstraintMessage(EvaluationContext ec, FormInstance instance,
     private void attemptConstraintCompile() {
         xPathConstraintMsg = null;
         try {
-            if(constraintMsg != null) {
+            if (constraintMsg != null && constraintMsg.startsWith(ITEXT_OPEN) && constraintMsg.endsWith(ITEXT_CLOSE)) {
```

Comment:
As I see from running in the debugger (before this change), 
- parsing a two-word string, “Your message”, results in an exception, `org.javarosa.xpath.parser.XPathSyntaxException: 
Couldn't understand the expression starting at this point: string(Your͎message)`. The exception is ignored, and `xPathConstraintMsg` remains `null`.

- parsing a one-word string, “Message”, is successful, and results in xPathConstraintMsg being set to `{func-expr:string,{{path-expr:rel,{{step:child,Message}}}}}`

For both of these strings, xPathConstraintMsg should remain `null`. The defect in the code is that plain strings (not inside itext) are being parsed unnecessarily, with the result that in at least this case of a single word, xPathConstraintMsg is set to something incorrect, when it should be null. Your correction is to avoid the parsing when itext is not used by searching the message for itext tokens.

If all that is correct, then let me ask:
- Is it necessary to look for ITEXT_CLOSE at the end?
- Might constraintMsg have leading or trailing whitespace or any other characters that might prevent the check for ITEXT_OPEN from working?
- For more robustness, did you consider parsing in all cases, and only using the result if an examination of the parse result shows that itext was used, such as is the case here?
```
{func-expr:string,{{func-expr:jr:itext,{{str:'/data/q1:constraintMsg'}}}}}
```

Reply:
> For both of these strings, xPathConstraintMsg should remain null. The defect in the code is that plain strings (not inside itext) are being parsed unnecessarily, with the result that in at least this case of a single word, xPathConstraintMsg is set to something incorrect, when it should be null. Your correction is to avoid the parsing when itext is not used by searching the message for itext tokens.

correct.

> Is it necessary to look for ITEXT_CLOSE at the end?

probably it's not necessary but it's not a problem as well.

>Might constraintMsg have leading or trailing whitespace or any other characters that might prevent the check for ITEXT_OPEN from working?

maybe but why would it be better? I think we should just avoid parsing.

> For more robustness, did you consider parsing in all cases, and only using the result if an examination of the parse result shows that itext was used, such as is the case here?

I think I can't check that easily.

Reply:
>>Might constraintMsg have leading or trailing whitespace or any other characters that might prevent the check for ITEXT_OPEN from working?

>maybe but why would it be better? I think we should just avoid parsing.

No, I’m saying what if the form has whitespace before the ITEXT_OPEN?

Reply:
Ahh I see you are right I just added `trim()` function to avoid such problems.

Reply:
OK, I’ll look for the trim when you push. Also, should we have a test for this?

Reply:
I think it's ready. Please review.